### DISCUSSION TITLE
How to deal with void* parameters to MagickWand API methods?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5743] [Development] [answered]
The function `MagickBooleanType MagickSetImageProfile(MagickWand *wand,
  const char *name,const void *profile,const size_t length)`
 takes a parameter `const void *profile` .

Then in the implementation it casts it to `(unsigned char *) profile` and assigns it to the current image of the magick wand:
`SetStringInfoDatum(profile_info,(unsigned char *) profile);`

I understand the code, but what I am confused about is this thing accepting a `const void*` and casting it to `unsigned char*`, I don't understand this part, what exactly is expected as a value of `profile` param? Is is a string? Or a list of positive-1byte-integers (unsigned char*) in range [0,255]?

I want to know what does it expect as a `profile`, I mean can you give me an example about a valid arg to the `profile` param? I want to know exactly what this method expects in order to map it to the correct type in dart.

Note:
 - this question applies to all methods that accept void* parameters, and then it casts these parameters to the same type every time.
- I know what is `void*` pointer, but why do you take a void* pointer when only 1 possible type (unsigned char*) is being handled

**Comment by urban-warrior** [comment, marked as answer]
Typically, `void*` is used to pass a buffer of raw data, however, you always need to first cast the pointer to interpret it.  Since SetStringInfo*() typically operated on strings, it was built around `unsigned char*`.  Signed/unsigned characters are typically utilized whenever their is a length associated with data.  Whereas, with void pointers, a length would not make sense given void is a general declaration that can hold any pointer type.

We'll correct the `const` issue.   The profile is `const void*` and therefore we would assume `const unsigned char*` when placing it in a StringInfo container.

**Reply by Haidar0096** [reply]
> Typically, `void*` is used to pass a buffer of raw data, however, you always need to first cast the pointer to interpret it. Since SetStringInfo*() typically operated on strings, it was built around `unsigned char*`. Signed/unsigned characters are typically utilized whenever their is a length associated with data. Whereas, with void pointers, a length would not make sense given void is a general declaration that can hold any pointer type.
> 
> We'll correct the `const` issue. The profile is `const void*` and therefore we would assume `const unsigned char*` when placing it in a StringInfo container.

First of all, thanks for your reply :)

If I understand correctly, by profile you mean metadata that describes an image ( [ICC](https://en.wikipedia.org/wiki/ICC_profile), [IPTC](https://iptc.org/standards/photo-metadata/iptc-standard/), others(https://www.technospot.net/blogs/photome-edits-and-export-exif-iptc-naa-and-icc-profile-data-from-images/)...)

But you didn't answer me yet,  if I want to call this function from dart language, what should I pass as a value of profile? Should I pass a string? a byte array (array of `unsigned char`) ?

I do understand the code of `MagickSetImageProfile`, but I don't have an idea what is the format of `profile` array.

Why I need it? Because I am writing a binding in the dart language that will call this c method, so I need to know how should I represent profile, whether using a byte array, or a string, or ???

Again, sorry if I'm asking alot but details like these are not always clear as they seem to be.

**Reply by urban-warrior** [reply]
A profile can be anything, its a blob, including random bytes.  Its the interpretation that imposes restrictions on the bytes based on the profile handle.  For example, set a profile to EXIF, ImageMagick will expect the blob to conform to the EXIF format specification.  Set the profile to RANDOM and the bytes could be gibberish because ImageMagick won't interpret it.

**Reply by Haidar0096** [reply]
@urban-warrior this is the answer I was looking for, so on the dart side, `profile` will be a List of bytes (each byte ranges from 0 -> 255), and this will be passed to `MagickSetImageProfile`'s `profile` param.

Thank you.
